Refactor Pulse gates (now Pulse schedules) page

abbycross opened the pull request:
Closes #46

abbycross commented:
This is nearly ready for review @nkanazawa1989 ! A couple things to note: I am confused about your note regarding ShiftFrequency and SetPhase (see below) -- I thought they would be equivalent to the other instructions listed, taking the format `shift_frequency` and `set_phase`, but are they different?

> Some pulse instructions are missing; [ShiftFrequency](https://qiskit.org/documentation/stubs/qiskit.pulse.instructions.ShiftFrequency.html#qiskit.pulse.instructions.ShiftFrequency) and [SetPhase](https://qiskit.org/documentation/stubs/qiskit.pulse.instructions.SetPhase.html#qiskit.pulse.instructions.SetPhase) are basic instructions that worth explaining. IBM backend doesn't support set phase, so it's also worth mentioning that supported instructions are backend dependent.

abbycross commented:
I took a first try at adding in the note about shift/set frequency/shift in this commit [here](https://github.com/Qiskit/documentation/pull/217/commits/90ac6e2ac18b644079b43aea24e497fcef3732c5).

abbycross commented:
To address the library deprecation, it was suggested to me that I change all .gaussian to .Gaussian, since that pulls from a different library that won't be deprecated. (It looks like eventually, when the naming conventions are synced back up post-deprecation, there will be a need to change them back to lower-case)
 Let me know @nkanazawa1989 if that isn't the right way to solve the problem.

abbycross commented:
I added (very anemic) brief descriptions of shift_frequency and set_phase, plus code examples from the API docs, [in this commit](https://github.com/Qiskit/documentation/pull/217/commits/500e24c34067851e1ca1bdc8ba53808cb515d9b9). These two sections would benefit from some additional info from a pulse expert!

javabster reviewed (commented):
This generally LGTM, but I'd like to get @nkanazawa1989 to approve before we merge this.

One small thing, when you say "This behavior is backend-dependent, and might be updated in future controllers" I think we should drop the bit about maybe updating things in future. If we're not entirely sure this will happen I don't think we should mention it to users so directly

abbycross commented on docs/build/pulse.ipynb:
```diff
@@ -770,6 +776,59 @@
+    "<Admonition type=\"note\" title=\"Note\">\n",
+    "The `shift_frequency`, `shift_phase`, `set_frequency`, and `set_phase` instructions are not only changing the frequency of following pulses - they also change the frame of reference of the channel. Because a qubit oscillates at the frequency of its transition, the controller needs to sync with its oscillation; otherwise, unwanted Z drive is continuously applied. Usually the frame is matched  ...
+    "\n",
+    "Note also that these instructions are localized in the pulse gate in IBM devices. This means that accumulated phase and frequency shifts are not carried over. Each pulse gate always starts from the hardware default setting. This behavior is backend-dependent, and might be updated in future controllers.\n",
```
```suggestion
    "Note also that these instructions are localized in the pulse gate in IBM devices. This means that accumulated phase and frequency shifts are not carried over. Each pulse gate always starts from the hardware default setting. This behavior is backend-dependent.\n",
```

nkanazawa1989 commented on docs/build/_toc.json:
```diff
@@ -47,7 +47,7 @@
       "children": [
         {
-          "title": "Pulse gates",
+          "title": "Tune pulse waveforms",
```
Tune pulse waveforms would be okey if this is ad headline like. Strictly speaking, pulse gate is a microcode that defines a calibration of gate, and what it provides is a schedule (pulse sequence) rather than waveform.

abbycross replied:
Hm, how about `Tune pulse schedules`? I think @kaelynj suggested "Tune pulse waveforms", so tagging her if she has more feedback!

kaelynj replied:
I think `Tune Pulse Schedules` or `Tune Pulse Instructions` would work well (also maybe "Tuning" instead of "Tune"?).  Mostly I'm thinking of trying to communicate a broad idea to the reader of what Pulse is used fore.

abbycross replied:
Tune is more active than Tuning, but another option would be to remove the verb altogether if it sounds too weird with it in. I got feedback at lunch today to drop Tune from the title. What do we think of just `Pulse schedules`?

javabster replied:
Let's go with Pulse Schedules :)

nkanazawa1989 commented on docs/build/pulse.ipynb:
```diff
@@ -74,9 +74,9 @@
     "Now that we have our circuit, let's define a calibration for the Hadamard gate on qubit 0.\n",
     "\n",
-    "In practice, the pulse shape and its parameters would be optimized through a series of Rabi experiments (see the [Qiskit Textbook](https://learn.qiskit.org/course/quantum-hardware-pulses/calibrating-qubits-using-qiskit-pulse) for a walkthrough). For this demonstration, our Hadamard will be a Gaussian pulse. We will _play_ our pulse on the _drive_ channel of qubit 0.\n",
+    "In practice, the pulse shape and its parameters would be optimized through a series of Rabi experiments. For this demonstration, our Hadamard will be a Gaussian pulse. We will _play_ our pulse on the _drive_ channel of qubit 0.\n",
```
```suggestion
    "In practice, the pulse shape and its parameters would be optimized through a series of calibration experiments. For this demonstration, our Hadamard will be a Gaussian pulse. We will _play_ our pulse on the _drive_ channel of qubit 0.\n",
```

abbycross replied:
Fixed [here](https://github.com/Qiskit/documentation/pull/217/commits/a2ea12da47525ae3afaf7b46d8ea8bcd55f59834)

nkanazawa1989 commented on docs/build/pulse.ipynb:
```diff
@@ -770,6 +776,59 @@
+   "metadata": {},
+   "source": [
+    "<Admonition type=\"note\" title=\"Note\">\n",
+    "The `shift_frequency`, `shift_phase`, `set_frequency`, and `set_phase` instructions are not only changing the frequency of following pulses - they also change the frame of reference of the channel. Because a qubit oscillates at the frequency of its transition, the controller needs to sync with its oscillation; otherwise, unwanted Z drive is continuously applied. Usually the frame is matched  ...
```
code directive seems like not working here. 
![image](https://github.com/Qiskit/documentation/assets/39517270/615cbbdf-ebfd-4a1d-a28e-699420da6f66)

abbycross replied:
Is this still broken on your end? Looks ok to me after [this commit](https://github.com/Qiskit/documentation/pull/217/commits/098faaefb0af3a0f60ffb232ee70dd8371b0c708).

nkanazawa1989 commented on docs/build/pulse.ipynb:
```diff
@@ -770,6 +776,59 @@
+   "metadata": {},
+   "source": [
+    "<Admonition type=\"note\" title=\"Note\">\n",
+    "The `shift_frequency`, `shift_phase`, `set_frequency`, and `set_phase` instructions are not only changing the frequency of following pulses - they also change the frame of reference of the channel. Because a qubit oscillates at the frequency of its transition, the controller needs to sync with its oscillation; otherwise, unwanted Z drive is continuously applied. Usually the frame is matched  ...
```
```suggestion
    "The `shift_frequency` and `set_frequency` instructions are not only changing the frequency of following pulses - they also change the frame of reference of the channel. Because a qubit oscillates at the frequency of its transition, the controller needs to sync with its oscillation; otherwise, unwanted Z drive is continuously applied. Usually the frame is matched with the frequency of the drive, and drive matches with the frequency of the transition; the Z drive is eliminated when the qubit frequency is calibrated properly. When you apply the `shift_frequency` instruction, it changes the drive frequency and impacts the frame; in other words, it accumulates the phase (Z) as a function of shifted frequency and duration of the program. Specifically, when you shift the frequency by `df` and spend `dt` on that frame, the qubit may experience a phase rotation of `df * dt`. The programmer needs to take this into account to control their qubits precisely. \n",
```

nkanazawa1989 commented on docs/build/pulse.ipynb:
```diff
@@ -779,7 +838,7 @@
     "The `acquire` instruction triggers data acquisition for readout. It takes a duration, an `AcquireChannel` which maps to the qubit being measured, and a `MemorySlot` or a `RegisterSlot`. The `MemorySlot` is classical memory where the readout result will be stored. The `RegisterSlot` maps to a register in the control electronics which stores the readout result for fast feedback.\n",
     "\n",
-    "The `acquire` instructions can also take custom `Discriminator`s and `Kernel`s as keyword arguments."
+    "The `acquire` instruction can also take custom `Discriminator`s and `Kernel`s as keyword arguments. If you are interested in changing the measurement pulse envelope, the `Kernel` subroutine integrates time series of measurement response and generates an IQ data point, which will be classified into a quantum state by the discriminator. This indicates that if a user uses custom measurement sti ...
```
```suggestion
    "The `acquire` instruction can also take custom `Discriminator`s and `Kernel`s as keyword arguments. The `Kernel` subroutine integrates time series of measurement response and generates an IQ data point, which will be classified into a quantum state by the discriminator. This indicates that if a user uses custom measurement stimulus, as in a measurement pulse, one may need to update the kernel setting not to deteriorate the measurement SNR."
```

abbycross replied:
Fixed in https://github.com/Qiskit/documentation/pull/217/commits/098faaefb0af3a0f60ffb232ee70dd8371b0c708

nkanazawa1989 reviewed (commented):
Github doesn't allow me to comment on unmodified lines, so let me put comments here. Overall the docs looks almost good to go. Just minor suggestions.

- I often see "logical" in the tutorial. Originally this word is used to represent some abstract operation, but recently this is much used by the QEC community. Pulse gate doesn't create logical quantum circuit -- I'm fine with it if other people feel okey.

- In ScheduleBlock Instruction > play > Pulses, "The image below may provide some intuition for why they are specified separately. ..." is left while the image is gone. We can delete entire paragraph including Note comment "The hardware may be implemented in other ways, ...". This document might confuse experienced users. Indeed Qiskit Pulse aims at hiding such hardware details.

nitpick:
- The ordering of set_frequency, shift_phase, shift_phase, shift_frequency seems bit randomized. Perhaps set_frequency, shift_frequency, set_phase, shift_phase?

abbycross commented:
Thank you so much @nkanazawa1989 for all your help!

> 
>     * I often see "logical" in the tutorial. Originally this word is used to represent some abstract operation, but recently this is much used by the QEC community. Pulse gate doesn't create logical quantum circuit -- I'm fine with it if other people feel okey.

I took a quick poll and the prevailing opinion is that it makes sense in context, particularly for the relevant audience. Leaving as-is for now. (Suggestions for replacement are "circuit-level" and "digital")
> 
>     * In ScheduleBlock Instruction > play > Pulses, "The image below may provide some intuition for why they are specified separately. ..." is left while the image is gone. We can delete entire paragraph including Note comment "The hardware may be implemented in other ways, ...". This document might confuse experienced users. Indeed Qiskit Pulse aims at hiding such hardware details.

Great catch, thank you! removed in https://github.com/Qiskit/documentation/pull/217/commits/3abf1d641b4611da3ca5f14308a58caa19c95f3d
> 
> 
> nitpick:
> 
>     * The ordering of set_frequency, shift_phase, shift_phase, shift_frequency seems bit randomized. Perhaps set_frequency, shift_frequency, set_phase, shift_phase?

It was indeed totally random :) Fixed in [this commit](https://github.com/Qiskit/documentation/pull/217/commits/2bd60973aa42bb9f5e4e8f9dd530928058d318d9), and I also moved the note closer to the set/shift frequency sections.

abbycross commented:
Unless there are further suggested changes, I think all we need is to settle on the title of the page :)

abbycross commented on docs/build/_toc.json:
```diff
@@ -47,7 +47,7 @@
       "children": [
         {
-          "title": "Pulse gates",
+          "title": "Tune pulse waveforms",
```
```suggestion
          "title": "Pulse schedules",
```

abbycross commented on docs/build/pulse.ipynb:
```diff
@@ -4,17 +4,17 @@
    "metadata": {},
    "source": [
-    "# Pulse gates\n",
+    "# Tune pulse waveforms\n",
```
```suggestion
    "# Pulse schedules\n",
```

abbycross commented on docs/build/pulse.ipynb:
```diff
@@ -994,12 +1049,18 @@
  ],
  "metadata": {
-  "description": "Pulse gates notebook",
+  "description": "Tune pulse waveforms",
```
```suggestion
  "description": "Pulse schedules",
```

javabster reviewed (approved):
LGTM 🚀

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -533,15 +570,11 @@
+    "A `Pulse` specifies an arbitrary pulse _envelope_. The modulation frequency and phase of the output waveform are controlled by the [`set_frequency`](#set_frequency) and [`shift_phase`](#shift_phase) instructions.\n",
     "\n",
-    "**Note**: The hardware may be implemented in other ways, but if we keep the instructions separate, we avoid losing explicit information, such as the value of the modulation frequency.\n",
+    "There are many methods available for building up pulses. The `library` within Qiskit Pulse contains helpful methods for building `Pulse`s. Take for example a simple Gaussian pulse -- a pulse with its envelope described by a sampled Gaussian function. We arbitrarily choose an amplitude of 1, standard deviation $\\sigma$ of 10, and 128 sample points.\n",
```
```suggestion
    "There are many methods available for building pulses, such as those available in the Qiskit Pulse `library`.  Take for example a simple Gaussian pulse -- a pulse with its envelope described by a sampled Gaussian function. We arbitrarily choose an amplitude of 1, standard deviation $\\sigma$ of 10, and 128 sample points.\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -533,15 +570,11 @@
-    "There are many methods available to us for building up pulses. Our `library` within Qiskit Pulse contains helpful methods for building `Pulse`s. Let's take for example a simple Gaussian pulse -- a pulse with its envelope described by a sampled Gaussian function. We arbitrarily choose an amplitude of 1, standard deviation $\\sigma$ of 10, and 128 sample points.\n",
-    "\n",
-    "**Note**: The amplitude norm is arbitrarily limited to `1.0`. Each backend system may also impose further constraints -- for instance, a minimum pulse size of 64. These additional constraints, if available, would be provided through the `BackendConfiguration` which is described [here](https://qiskit.org/documentation/tutorials/circuits_advanced/08_gathering_system_information.html#Configurat ...
+    "**Note**: The amplitude norm is arbitrarily limited to `1.0`. Each backend system may also impose further constraints -- for instance, a minimum pulse size of 64. These additional constraints, if available, would be provided through `Target`, which is described [here](../api/qiskit/qiskit.transpiler.Target)."
```
```suggestion
    "**Note**: The amplitude norm is arbitrarily limited to `1.0`. Each backend system may also impose further constraints.  For instance, a minimum pulse size of 64. Any additional constraints are provided through [Target.](../api/qiskit/qiskit.transpiler.Target)"
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -563,9 +596,9 @@
    "source": [
     "##### Parametric pulses\n",
-    "Let's build our Gaussian pulse using the `Gaussian` parametric pulse. A parametric pulse sends the name of the function and its parameters to the backend, rather than every individual sample. Using parametric pulses makes the jobs you send to the backend much smaller. IBM Quantum backends limit the maximum job size that they accept, so parametric pulses may allow you to run larger programs.\ ...
+    "You can build your Gaussian pulse using the `Gaussian` parametric pulse. A parametric pulse sends the name of the function and its parameters to the backend, rather than every individual sample. Using parametric pulses makes the jobs you send to the backend much smaller. IBM Quantum backends limit the maximum job size that they accept, so parametric pulses may allow you to run larger program ...
```
```suggestion
    "You can build a Gaussian pulse by using the `Gaussian` parametric pulse. A parametric pulse sends the name of the function and its parameters to the backend, rather than every individual sample. Using parametric pulses makes the jobs much smaller to send. IBM Quantum backends limit the maximum job size that they accept, so parametric pulses might allow you to run larger programs.\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -563,9 +596,9 @@
+    "You can build your Gaussian pulse using the `Gaussian` parametric pulse. A parametric pulse sends the name of the function and its parameters to the backend, rather than every individual sample. Using parametric pulses makes the jobs you send to the backend much smaller. IBM Quantum backends limit the maximum job size that they accept, so parametric pulses may allow you to run larger program ...
     "\n",
-    "Other parametric pulses in the `library` include `GaussianSquare`, `Drag`, and `Constant`.\n",
+    "Other parametric pulses in the `library` include `GaussianSquare`, `Drag`, and `Constant`, among others. See the [full list in the API reference](../api/qiskit/pulse#parametric-pulse-representation).\n",
```
```suggestion
    "Other parametric pulses in the `library` include `GaussianSquare`, `Drag`, and `Constant`. See the [full list in the API reference](../api/qiskit/pulse#parametric-pulse-representation).\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -563,9 +596,9 @@
+    "Other parametric pulses in the `library` include `GaussianSquare`, `Drag`, and `Constant`, among others. See the [full list in the API reference](../api/qiskit/pulse#parametric-pulse-representation).\n",
     "\n",
     "\n",
     "**Note**: The backend is responsible for deciding exactly how to sample the parametric pulses. It is possible to draw parametric pulses, but the samples displayed are not guaranteed to be the same as those executed on the backend."
```
```suggestion
    "**Note**: The backend is responsible for deciding how to sample the parametric pulses. It is possible to draw parametric pulses, but the samples displayed are not guaranteed to be the same as those executed on the backend."
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -600,7 +633,7 @@
     "##### Pulse waveforms described by samples\n",
     "\n",
-    "A `Waveform` is a pulse signal specified as an array of time-ordered complex amplitudes, or _samples_. Each sample is played for one cycle, a timestep `dt`, determined by the backend. If we want to know the real-time dynamics of our program, we need to know the value of `dt`. The (zero-indexed) $i^{th}$ sample will play from time `i*dt` up to `(i + 1)*dt`, modulated by the qubit frequency."
+    "A `Waveform` is a pulse signal specified as an array of time-ordered complex amplitudes, or _samples_. Each sample is played for one cycle, a timestep `dt`, determined by the backend. Knowing the value of `dt` is necessary in determining the real-time dynamics of a program. The (zero-indexed) $i^{th}$ sample will play from time `i*dt` up to `(i + 1)*dt`, modulated by the qubit frequency."
```
```suggestion
    "A `Waveform` is a pulse signal specified as an array of time-ordered complex amplitudes, or _samples_. Each sample is played for one cycle, a timestep `dt`, determined by the backend. You must know the value of `dt` to determine a program's real-time dynamics. The (zero-indexed) $i^{th}$ sample plays from time `i*dt` up to `(i + 1)*dt`, modulated by the qubit frequency."
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -729,9 +731,9 @@
+    "As explained previously, the output pulse waveform envelope is also modulated by a frequency and phase. Each channel has a default frequency listed in the `backend.defaults`.\n",
     "\n",
-    "The frequency of a channel can be updated at any time within a `Schedule` by the `set_frequency` instruction. It takes a float `frequency` and a `PulseChannel` `channel` as input. All pulses on a channel following a `set_frequency` instruction will be modulated by the given frequency until another `set_frequency` instruction is encountered or until the program ends.\n",
+    "The frequency of a channel can be updated at any time within a `ScheduleBlock` by the `set_frequency` instruction. It takes a float `frequency` and a `PulseChannel` `channel` as input. All pulses on a channel following a `set_frequency` instruction will be modulated by the given frequency until another `set_frequency` instruction is encountered or until the program ends.\n",
```
```suggestion
    "A channel's frequency can be updated at any time within a `ScheduleBlock` by the `set_frequency` instruction. It takes a float `frequency` and a `PulseChannel` `channel` as input. All pulses on a channel following a `set_frequency` instruction are modulated by the given frequency until another `set_frequency` instruction is encountered or until the program ends.\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -748,6 +750,59 @@
+   "metadata": {},
+   "source": [
+    "<Admonition type=\"note\" title=\"Note\">\n",
+    "The `shift_frequency` and `set_frequency` instructions are not only changing the frequency of following pulses - they also change the frame of reference of the channel. Because a qubit oscillates at the frequency of its transition, the controller needs to sync with its oscillation; otherwise, unwanted Z drive is continuously applied. Usually the frame is matched with the frequency of the dri ...
```
```suggestion
    "The `shift_frequency` and `set_frequency` instructions change the frequency of following pulses and also change the channel's frame of reference. Because a qubit oscillates at its transition frequency, the controller needs to sync with its oscillation; otherwise, unwanted Z drive is continuously applied. Usually, because the frame is matched with the drive's frequency, and drive matches with the transition's frequency; the Z drive is eliminated when the qubit frequency is calibrated properly. When you apply the `shift_frequency` instruction, it changes the drive frequency and impacts the frame. In other words, it accumulates the phase (Z) as a function of shifted frequency and duration of the program. Specifically, when you shift the frequency by `df` and spend `dt` on that frame, the qubit may experience a phase rotation of `df * dt`. The programmer needs to take this into account to control their qubits precisely. \n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -779,7 +834,7 @@
     "\n",
     "The `acquire` instruction triggers data acquisition for readout. It takes a duration, an `AcquireChannel` which maps to the qubit being measured, and a `MemorySlot` or a `RegisterSlot`. The `MemorySlot` is classical memory where the readout result will be stored. The `RegisterSlot` maps to a register in the control electronics which stores the readout result for fast feedback.\n",
```
```suggestion
    "The `acquire` instruction triggers data acquisition for readout. It takes a duration, an `AcquireChannel`, which maps to the qubit being measured, and a `MemorySlot` or a `RegisterSlot`. The `MemorySlot` is classical memory where the readout result will be stored. The `RegisterSlot` maps to a register in the control electronics that stores the readout result for fast feedback.\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -779,7 +834,7 @@
     "The `acquire` instruction triggers data acquisition for readout. It takes a duration, an `AcquireChannel` which maps to the qubit being measured, and a `MemorySlot` or a `RegisterSlot`. The `MemorySlot` is classical memory where the readout result will be stored. The `RegisterSlot` maps to a register in the control electronics which stores the readout result for fast feedback.\n",
     "\n",
-    "The `acquire` instructions can also take custom `Discriminator`s and `Kernel`s as keyword arguments."
+    "The `acquire` instruction can also take custom `Discriminator`s and `Kernel`s as keyword arguments. The `Kernel` subroutine integrates time series of measurement response and generates an IQ data point, which will be classified into a quantum state by the discriminator. This indicates that if you use custom measurement stimulus, as in a measurement pulse, you may need to update the kernel se ...
```
```suggestion
    "The `acquire` instruction can also take custom `Discriminator`s and `Kernel`s as keyword arguments. The `Kernel` subroutine integrates a time series of measurement responses and generates an IQ data point, which will be classified into a quantum state by the discriminator. This indicates that if you use a custom measurement stimulus, as in a measurement pulse, you might need to update the kernel setting to not deteriorate the measurement SNR."
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -799,15 +854,15 @@
    "metadata": {},
    "source": [
-    "Now that we know how to add `Schedule` instructions, let's learn how to control exactly when they're played.\n",
+    "You now know how to add `ScheduleBlock` instructions. Next, learn how to control exactly when they're played.\n",
```
```suggestion
    "After adding `ScheduleBlock` instructions, you need to understand how to control when they're played.\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -799,15 +854,15 @@
     "\n",
     "## Pulse Builder\n",
-    "Here, we will go over the most important Pulse Builder features for learning how to build schedules. This is not exhaustive; for more details about what you can do using the Pulse Builder, check out the [Pulse API reference](/api/qiskit/pulse).\n",
+    "Below are the most important Pulse Builder features for learning how to build schedules. This is not exhaustive; for more details about what you can do using the Pulse Builder, check out the [Pulse API reference](/api/qiskit/pulse).\n",
```
```suggestion
    "Below are the most important Pulse Builder features for learning how to build schedules. This is not an exhaustive list.  For more details about using the Pulse Builder, refer to the [Pulse API reference.](/api/qiskit/pulse)\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -799,15 +854,15 @@
+    "Below are the most important Pulse Builder features for learning how to build schedules. This is not exhaustive; for more details about what you can do using the Pulse Builder, check out the [Pulse API reference](/api/qiskit/pulse).\n",
     "\n",
     "### Alignment contexts\n",
     "The builder has alignment contexts which influence how a schedule is built. Contexts can also be nested. Try them out, and use `.draw()` to see how the pulses are aligned.\n",
```
```suggestion
    "The builder has alignment contexts that influence how a schedule is built. Contexts can also be nested. Try them out, and use `.draw()` to see how the pulses are aligned.\n",
```

beckykd commented on docs/build/pulse.ipynb:
```diff
@@ -849,7 +904,7 @@
     "\n",
     "### `align_right`\n",
-    "Unsurprisingly, `align_right` does the opposite of `align_left`. It will choose `t=100` in the above example to begin the gaussian pulse on `D0`. Left and right are also sometimes called \"as soon as possible\" and \"as late as possible\" scheduling, respectively."
+    "Unsurprisingly, `align_right` does the opposite of `align_left`. It will choose `t=100` in the above example to begin the Gaussian pulse on `D0`. Left and right are also sometimes called \"as soon as possible\" and \"as late as possible\" scheduling, respectively."
```
```suggestion
    "`align_right` does the opposite of `align_left`. It chooses `t=100` in the above example to begin the Gaussian pulse on `D0`. Left and right are also sometimes called \"as soon as possible\" and \"as late as possible\" scheduling, respectively."
```
